docs/RFCs/19042021_Cert_Management_In_CRDB_Helm_Chart.md

```diff
@@ -0,0 +1,243 @@
+  * The schedule of the cronjob will be of minimum `clientCertDuration` and `nodeCertDuration` minus some delta.
+  * On every scheduled run, it will check if there is any certificate that is going to expire before the next scheduled run,
+   if yes then it will renew the certificates.
+  
+  * <b>The cronjob will use the same `pre-install` job image for certificate rotations. The `pre-install` job image binary will have an argument `--rotate` for handling certificate rotation.</b>
+ 
+* The Stateful pod will be changed to only run `copy-certs` initContainer to copy the certificates from Node secret to emptyDir volume.
+  The rest of the main DB container flow will remain the same.
+* The post-install job will be changed to only run `copy-certs` initContainer to copy the certificates from Client secret to emptyDir volume.
+  The rest of the main cluster-init container flow will remain the same.
+ 
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
+ 
+### Certificate Management Service Implementation
+ 
+ 
+### Overall Flow
+ 
+* Check if CA secret is empty
+  * if yes, [Generate-CA](#generate-ca), [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert); return
+  * if not, [Validate Secret Annotations for CA](#validate-secret-annotations)
+    * if valid, CA is intact
+    * if not, [Annotate-CA](#annotate-ca), [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert); return
+* Check if [CA requires rotation](#check-cert-for-regeneration)
+   * if yes, follow [Rotate-CA](#rotate-ca); return
+* Check if Node or Client certificates [needs to be regenerated](#check-cert-for-regeneration):
+   * if yes, follow [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert)
+ 
+ 
+ 
+#### Generate-CA
+* A self-signed CA will be generated
+* Expiry of the certificate will be driven by the Helm value for `tls.certs.generate.caCertDuration`, passed as env variable
+* Contents of the CA certificate will be stored in the default CA secret `cockroachdb-ca`.
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Follow [CA Annotation Workflow](#annotate-ca)
+ 
+#### Generate-Node-Cert
+* A Node certificate will be generated signed by the generated CA or user custom CA.
+* Expiry of the certificate will be driven by the Helm value for `tls.certs.generate.nodeCertDuration`, passed as env variable
+* Contents of the Node certificate will be stored in the Node secret `cockroachdb-node`.
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Node secret will be patched with annotation `resourceVersion` with the value of its current resourceVersion
+* Node secret will be patched with annotation `creationTime` and `duration` with current UTC time and value from `tls.certs.generate.nodeCertDuration`
+ 
+#### Generate-Client-Cert
```

Comment:
We will create a helm plugin that can be used to generate any user's certificate apart from the root user certificate. Once a new user's certificate is created, it will be rotated/handled by the existing cert-rotation cronjobs.